q2_types/per_sample_sequences/_format.py

```diff
@@ -137,21 +142,56 @@ def sequences_path_maker(self, sample_id, barcode_id, lane_number,
         return '%s_%s_L%03d_R%d_001.fastq.gz' % (sample_id, barcode_id,
                                                  lane_number, read_number)
 
+    def _find_duplicates(self, ids):
+        return {x for x, c in collections.Counter(ids).items() if c > 1}
+
+    def validate(self):
+        forwards = []
+        reverse = []
+        for p in self.path.iterdir():
+            if p.is_dir():
+                d = p.relative_to(self.path)
+                raise ValidationError("Contains a subdirectory: %s" % d)
+            else:
+                if p.name.endswith('_001.fastq.gz'):
+                    sample_id = p.name.rsplit('_', maxsplit=4)[0]
+                    if p.name.endswith('R1_001.fastq.gz'):
+                        forwards.append(sample_id)
+                    else:
+                        reverse.append(sample_id)
+
+        set_forwards = set(forwards)
+        set_reverse = set(reverse)
+
+        if len(set_forwards) != len(forwards):
+            raise ValidationError('Duplicate samples in forward reads: %r'
+                                  % self._find_duplicates(forwards))
+        if len(set_reverse) != len(reverse):
+            raise ValidationError('Duplicate samples in reverse reads: %r'
+                                  % self._find_duplicates(reverse))
+
+        if forwards and reverse:
+            if not self._CHECK_PAIRED:
+                raise ValidationError("Forward and reverse reads found.")
+            elif set_forwards ^ set_reverse:
+                raise ValidationError(
+                    "These samples do not have matching pairs of forward and "
+                    "reverse reads: %r" % (set_forwards ^ set_reverse))
+        elif self._REQUIRE_PAIRED:
+            raise ValidationError("Reads are not paired end.")
```

Comment:
This error message will trigger if there aren't any forward reads, so you might want to have a different error message for that case (there should always be at least one file of forward reads).

Reply:
I thought we permitted `R2` files to be imported as single-end (so they could all be uniformly your reverse reads), that's why the format names don't use the word "forward", just "single".

Reply:
Makes sense, thanks!